File: todo/todo.go
package todo

import (
	"go-todo-web/category"
	"go-todo-web/database"

	"gorm.io/gorm"
)

var db *gorm.DB = database.GetDatabase()

// todo struct
type Todo struct {
	gorm.Model
	Name       string `json:"name"`
	CategoryId uint   `json:"category_id"`
}

/*
	Create todo
	@param name string
*/
func CreateTodo(name string, categoryId uint) uint {
	// check if category is valid
	if category.CheckValidCategoryId(categoryId) == false {
		panic("category id is not valid")
	}

	// create new todo
	todo := Todo{Name: name, CategoryId: categoryId}

	// save todo
	db.Create(&todo)

	// return id
	return todo.ID
}

/*
	Update todo
	@param id int
	@param name string
*/
func UpdateTodo(id int, name string, categoryId uint) {
	// check if category is valid
	if category.CheckValidCategoryId(categoryId) == false {
		panic("category id is not valid")
	}

	// update todo
	db.Model(&Todo{}).Where("id = ?", id).Update("name", name).Update("category_id", categoryId)
}

/*
	Delete todo
	@param id int
*/
func DeleteTodo(id int) {
	// delete todo
	db.Delete(&Todo{}, id)
}

File: frontend/src/api/todo.ts
import { API } from "./api";

export interface ITodo {
  id: number;
  category_id: number;
  name: string;
}

export class TodoAPI extends API {
  /**
   * getTodo
   * @returns {Promise<ITodo[]>}
   */
  public static async getTodo(): Promise<ITodo[]> {
    const request = await fetch(this.getAPIHost() + "/api/todo");
    return request.json();
  }

  /**
   * addTodo
   * @param todoName
   * @param categoryId
   * @returns {Promise<void>}
   */
  public static async addTodo(
    todoName: string,
    categoryId: number
  ): Promise<void> {
    await fetch(this.getAPIHost() + "/api/todo/create", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ name: todoName, category_id: categoryId }),
    });
  }

  /**
   * updateTodo
   * @param todoId
   * @param todoName
   * @param categoryId
   * @returns {Promise<void>}
   */
  public static async updateTodo(
    todoId: number,
    todoName: string,
    categoryId: number
  ): Promise<void> {
    await fetch(this.getAPIHost() + "/api/todo/update", {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        id: todoId,
        name: todoName,
        category_id: categoryId,
      }),
    });
  }

  /**
   * deleteTodo
   * @param todoId
   * @returns {Promise<void>}
   */
  public static async deleteTodo(todoId: number): Promise<void> {
    await fetch(this.getAPIHost() + "/api/todo/delete", {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        id: todoId,
      }),
    });
  }
}
